### ISSUE TITLE
Inconsistent compile behavior for out parameters

### ISSUE THREAD

**Issue by Petermarcu** [dotnet/corefx#28392]
@marce155 commented on [Thu Mar 22 2018](https://github.com/dotnet/core/issues/1373)

When declaring an out parameter on a method I’m used to having to assign the parameter a value before leaving the method: 
```
public bool TryGet(string foo, out TimeSpan bar){
 bar = TimeSpan.FromSeconds(1);
 return true;
}
```

Today I encountered a behavior which was at least unexpected and feels inconsistent. When I’m targeting netcoreapp2.0 there are some cases where the program will compile and work even if the out parameter has not been assigned.

To be exact this is the case for TimeSpan or structs containing (only) a TimeSpan (that is as far as I tested). The same code will fail to compile for .NET 4.6.

```
class Program
    {
        public static void Main()
        {
            TimeSpan timeSpan;
            Foo foo;
            Bar bar;
            //Baz baz;

            TryTimeSpan(out timeSpan);
            TryFoo(out foo);
            TryBar(out bar);
            //TryBaz(out baz);

            Console.WriteLine(timeSpan);
            Console.WriteLine(foo);
            Console.WriteLine(bar);
        }
        private static bool TryTimeSpan(out TimeSpan timeSpan)
        {
	    // no compile error - unexpected
            return false;
        }
        private static bool TryFoo(out Foo foo)
        {
	    // empty struct works with target net461 as well
            return false;
        }
        private static bool TryBar(out Bar bar)
        {
	    // no compile error - unexpected
            return false;
        }
        /*
        private static bool TryBaz(out Baz baz)
        {
            // compile error, baz not assigned for all targets - expected
            return false;
        }
        */
    }
    struct Foo { }
    struct Bar
    {
        public TimeSpan TimeSpan { get; set; }
    }
    struct Baz
    {
        public int Int { get; set; }
    }
```
I assume that somehow proper defaults are assigned, yet it still feels a little strange. Could you please clarify if this is expected behavior and what causes it?

---

@danstur commented on [Thu Mar 22 2018](https://github.com/dotnet/core/issues/1373#issuecomment-375337998)

This definitely seems to violate the [specification](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#output-parameters): 

> Every output parameter of a method must be definitely assigned before the method returns.

**Comment by stephentoub** [comment]
Dup of https://github.com/dotnet/corefx/issues/6185
cc: @weshaggard

**Comment by weshaggard** [comment]
Yes this was an issue in the ref assemblies that is getting addressed this release with #6185.

**Comment by martinsuchan** [comment]
Just a note, the recently released Microsoft.NETCore.UniversalWindowsPlatform NuGet 6.2.2 has this issue - Windows Runtime structs cannot be used according to C# specification: https://github.com/Microsoft/dotnet/issues/927

![48697530-1eb4be80-ebe5-11e8-8e3a-d1bca379fc4d](https://user-images.githubusercontent.com/3041397/48761682-89c7c900-eca9-11e8-85b4-c9bd72179dd1.png)
